### DISCUSSION TITLE
Your account is inactive and marked for removal (Or is it? Turns out we can't be sure.)

### DISCUSSION THREAD

**Discussion by joostdecock** [freesewing/freesewing#4698] [Announcements]
Dear all,

Yesterday, I got contacted by a user who received an email like this:

> ## Your account is inactive and marked for removal
>
> **Your FreeSewing account is inactive and marked for removal**
>
> Due to inactivity, your account on FreeSewing.org has been marked for removal. 
> If you want to keep your account, please login on FreeSewing.org within the next 2 weeks.                                                                             > 
> 
> **Go to FreeSewing.org   →**
>
> If your account stays dormant, it will at one point become unavailable.
> That does not have to be a disaster, you can always create a new one, but you will lose all data stored in your account.

Problem is that the user in question had not been inactive. 
Much to the contrary, they use the site very often.

After investigation, we were able to confirm that there is a bug in the v2 backend that will not update the `lastLogin` date when an access token is refreshed prior to expiry. This sort of refresh happens transparently when you stay active on the site.

While the date is updated when you do an actual login, if you visit the site often enough that the token never expires (and gets refreshed on each visit) your `lastLogin` date is never updated.

**What makes a login a login?**

If we were to go ahead and cull these dormant accounts, we risk getting some of our most active users getting caught in the crossfire as false positives.
On the other hand, if we do not cull these dormant accounts, an argument can be made that it is in violation with our privacy notice that states that:

> *We keep profile data up to 12 months after your last login, or until you remove it.*

The word **login** is doing a lot of the heavy lifting here. What does it mean to *login* to the site?
If you visit the site again after two weeks and it transparently extends your authentication token, is that a login?
Or is it only a login when you re-enter your username and password and click the log in button?

This is a not a legal discussion, so let's instead look at the practical implications of both options:

- Token refresh is not a login. Data from accounts with a `lastLogin` date > 12 months should be removed. And if that means we remove data from our most active users, so be it.
- Token refresh is a login. Data from accounts with a `lastLogin` date > 12 months should not be removed. And if that means some actually dormant accounts don't get removed, so be it.

Both are not ideal. But one is a lot less ideal.

I am sorry for not realizing sooner this was an issue, and I apologize to those people who made the assumption that not using the site for a whole year would mean their data would be removed. Which, to be clear, is an absolutely reasonable assumption.
Alas, we can't seem to make good on that promise without inflicting some real damage on other users.

So, given that:
- All your data is encrypted at rest
- We do not share your data
- We do not sell your data

I am going with the lesser evil here and decide that **no data will be removed from any activated v2 user accounts**.

In other words: If you ever used your account, we will migrate it to v3.
But I will also use this as a lesson and make some changes to improve things going forward. Specifically:

- Add a `lastSeen` field to the database to be less ambiguous
- Ensure this `lastSeen` field is updated each time a user checks in
- Update our privacy notice to clarify the phrasing so it no longer hinges on the interpretation of what constitutes *a login*.
- Migrate all current (active) v2 accounts to v3

Obviously, all users can at any time [remove their account](https://freesewing.org/account/actions/remove/) that has not changed.

Once again, please forgive me for the confusion, and if you received one of the emails above, rest assured **your account will not be removed**.

love,
joost

**Comment by LeviFiction** [comment]
Interesting, I thought most software just put a timeout date on a token and
only refreshed it on an active login.  Never occurred to me a site would
auto renew tokens as you use the site.  Interesting problem to have.

**Comment by joostdecock** [comment]
- Backend changes implemented in #4717
- Privacy notice updated in #4718
